omniqueue/src/backends/sqs.rs

```diff
@@ -305,10 +308,80 @@ impl SqsProducer {
         let payload = serde_json::to_string(payload)?;
         self.send_raw_scheduled(&payload, delay).await
     }
+
+    #[tracing::instrument(name = "send_batch", skip_all)]
+    async fn send_batch_inner<I>(
+        &self,
+        payloads: impl IntoIterator<Item = I, IntoIter: Send> + Send,
+        convert_payload: impl Fn(I) -> Result<String>,
+    ) -> Result<()> {
+        // Convert payloads up front and collect to Vec to run the payload size
+        // check on everything before submitting the first batch.
+        let payloads: Vec<_> = payloads
+            .into_iter()
+            .map(convert_payload)
+            .collect::<Result<_>>()?;
+
+        for payload in &payloads {
+            if payload.len() > MAX_PAYLOAD_SIZE {
+                return Err(QueueError::PayloadTooLarge {
+                    limit: MAX_PAYLOAD_SIZE,
+                    actual: payload.len(),
+                });
+            }
+        }
+
+        for payloads in payloads.chunks(MAX_BATCH_SIZE) {
+            let entries = payloads
+                .iter()
+                .enumerate()
+                .map(|(i, payload)| {
+                    SendMessageBatchRequestEntry::builder()
+                        .message_body(payload)
+                        .id(i.to_string())
+                        .build()
+                        .map_err(QueueError::generic)
+                })
+                .collect::<Result<_>>()?;
+
+            self.client
+                .send_message_batch()
+                .queue_url(&self.queue_dsn)
+                .set_entries(Some(entries))
+                .send()
+                .await
+                .map_err(aws_to_queue_error)?;
+        }
+
+        Ok(())
+    }
 }
 
-impl_queue_producer!(SqsProducer, String);
-impl_scheduled_queue_producer!(SqsProducer, String);
+impl crate::QueueProducer for SqsProducer {
+    type Payload = String;
+    omni_delegate!(send_raw, send_serde_json);
+
+    /// This method is overwritten for the SQS backend to be more efficient
+    /// than the default of sequentially publishing `payloads`.
+    fn send_raw_batch(
+        &self,
+        payloads: impl IntoIterator<Item: AsRef<Self::Payload> + Send, IntoIter: Send> + Send,
+    ) -> impl Future<Output = Result<()>> {
+        self.send_batch_inner(payloads, |p| Ok(p.as_ref().into()))
+    }
+
+    /// This method is overwritten for the SQS backend to be more efficient
+    /// than the default of sequentially publishing `payloads`.
+    fn send_serde_json_batch(
```

Comment:
Overall, it looks good to me. The only thing I find weird about the API is that users of `send_serde_json_batch` kind of need to know how it works internally (`serde_json::to_string`) to get the size of their payload.

Reply:
But I guess that also happens with non-batch methods

Reply:
Yeah, can always pre-serialize and use one of the other methods if needed.